Opmon protobuf microservice

mroda88 opened the pull request:
This PR creates a new microservice that uses the new protobuf opmon stream and puts it into influxdb. Eventually this will replace the old stream, but for now this only deploys the new one.

mroda88 committed:
Initial file for the opmon microservice

mroda88 committed:
initial draft for testing

mroda88 committed:
Running state for tests

mroda88 committed:
correct creation of the json structure

mroda88 committed:
a bit of cleanup

mroda88 committed:
Great progress in the microservice

mroda88 committed:
Some clenup

mroda88 committed:
Passing the client to the consumer

mroda88 committed:
Tentative entry file

mroda88 committed:
Draft of yaml file

mroda88 committed:
Update dependencies

mroda88 committed:
Implement suggestions from Pat

mroda88 committed:
Correct dependencies

mroda88 committed:
Merge remote-tracking branch 'origin/develop' into mroda/opmon_protobuf

mroda88 committed:
fix typo

mroda88 committed:
Buildable dependency

mroda88 committed:
Remove debug printouts

mroda88 committed:
add the right image for documentation

jcpunk commented on opmon-protobuf-dbwriter/dbwriter.py:
```diff
@@ -0,0 +1,168 @@
+@click.option('--subscriber-topic',     type=click.STRING, multiple=True, default=['opmon_stream'], help='The system will add the "monitoring." prefix' )
+
+#influx options
+@click.option('--influxdb-address', type=click.STRING, default='np04-srv-017', help='address of the influx db')
```
Should this be `np04-srv-017.cern.ch`?

mroda88 replied:
In fact I came put monkafka.cern.ch so that we make things more uniform to the other microservices.

jcpunk commented:
I'd be tempted to run the python code through the `black` autoformatter.

jcpunk commented on opmon-protobuf-dbwriter/dbwriter.py:
```diff
@@ -0,0 +1,168 @@
+                batch_ms = entry.ms
+            
+        except queue.Empty :
+            logging.debug("Queue is empty")
```
I'd be tempted to make this log at `info` level.  But you know the code/workflow better than I.  If you think `debug` is the right level that is good enough for me.

mroda88 replied:
I think this is ok otherwise this will create a huge log and for nothing really.

jcpunk commented on opmon-protobuf-dbwriter/dbwriter.py:
```diff
@@ -0,0 +1,168 @@
+#influx options
+@click.option('--influxdb-address', type=click.STRING, default='np04-srv-017', help='address of the influx db')
+@click.option('--influxdb-port',    type=click.INT,    default=31002,          help='port of the influxdb')
+@click.option('--influxdb-name',    type=click.STRING, default='test_influx',  help='Table name destination inside influxdb')
```
Do we need a username/password also to permit write access?

mroda88 replied:
Is it a question for the microservice or for influx itself? At the moment this code should work. But if we need to add name and password we can probably add it.

jcpunk replied:
I don't think we've got influxdb configured for unauthenticated writes.

mroda88 replied:
Exactly, that's why this code will work as it is. Do we think it's a problem and do we want to fix it now? Or can we leave it for after the whole HD run?

jcpunk replied:
I'd be a bit surprised if the service was able to write to influxdb without the user/pass set.  I think args for those will need to be added.

mroda88 replied:
The old one did. But I agree that we can add them so the image is done once and for all. I'll do it first thing tomorrow.

mroda88 committed:
black format and update in the servers

mroda88 commented:
Format with black is done

jcfreeman2 commented:
Basic code structure in `opmon-protobuf-dbwriter/` looks correct. There's a trailing whitespace in the string passed to `opmon-protobuf-dbwriter/entrypoint.sh`'s `ensure_required_variables`, but it shouldn't matter.

jcfreeman2 commented:
If I want to actually try running this microservice, what are some reasonable values to set `OPMON_DBWRITER_KAFKA_GROUP `, etc., to? And are these documented anywhere?

Possibly relevant here is Issue #104

mroda88 commented:
Hi John, so good values for this micro service are set in the yaml file in the same directory. Those are working values based on previous tests with opmon and influx. 

But running it effectively outside the kubernetes cluster is impossible. The influx server is configured in such a way that it has to be contacted from within the cluster. That's why for now this is the initial PR only and the test will be done once beam is off next week and we can tamper with the cluster freely. 

But if you want to, we can sit down tomorrow and see what we can do.

mroda88 commented:
I'll update the readme, thanks for pointing it out.

jcfreeman2 commented:
Thanks! Also FWIW, if I try this at the command line:
```
docker run --rm -e MICROSERVICE=opmon-protobuf-dbwriter -e OPMON_DBWRITER_KAFKA_BOOTSTRAP_SERVER=dummyval -e OPMON_DBWRITER_KAFKA_GROUP=dummyval -e OPMON_DBWRITER_SUBSCRIBER_TIMEOUT_MS=dummyval -e OPMON_DBWRITER_SUBSCRIBER_TIMEOUT_MS=dummyval -e OPMON_DBWRITER_TOPIC=dummyval -e OPMON_DBWRITER_INFLUX_HOST=dummyval -e OPMON_DBWRITER_INFLUX_PORT=dummyval -e OPMON_DBWRITER_TABLE=dummyval -e OPMON_DBWRITER_BATCH_SIZE_MS=dummyval ghcr.io/dune-daq/microservices:mroda-opmon_protobuf
```
...then I see
```
Usage: dbwriter.py [OPTIONS]
Try 'dbwriter.py -h' for help.

Error: No such option: --influx-name (Possible options: --influxdb-address, --influxdb-name, --influxdb-timeout)
Return value of call to /microservices/opmon-protobuf-dbwriter/entrypoint.sh is 2
```
where the error message appears unrelated to the obviously-phony values I provided the environment variables.

mroda88 commented:
Oh, it's probably because I put an empty line in the options. Tomorrow I'll sort them all since we are going to add user and password as well according to Pat's suggestion.

mroda88 committed:
Fix dbnmae in entrypoint

mroda88 committed:
Add username and password

mroda88 committed:
Balck format

mroda88 committed:
Correct inputs

mroda88 committed:
Add documentation

mroda88 commented:
ok, I think I addressed all the comments. Now if you do something like
```bash
docker run --rm -e MICROSERVICE=opmon-protobuf-dbwriter -e OPMON_DBWRITER_KAFKA_BOOTSTRAP_SERVER=dummyval -e OPMON_DBWRITER_KAFKA_GROUP=dummyval -e OPMON_DBWRITER_SUBSCRIBER_TIMEOUT_MS=10000  -e OPMON_DBWRITER_TOPIC=dummyval -e OPMON_DBWRITER_INFLUX_HOST=dummyval -e OPMON_DBWRITER_INFLUX_PORT=12345 -e OPMON_DBWRITER_TABLE=dummyval -e OPMON_DBWRITER_BATCH_SIZE_MS=1000 ghcr.io/dune-daq/microservices:mroda-opmon_protobuf
```
You get the correct failures due to connecting issues to the database.

mroda88 committed:
Correct indentation

jcpunk commented:
Cool! Do you have a snippet I could use to create the missing database?

mroda88 commented:
Ah, but that's the nice thing. This micro service will create the database if not already available. So we don't have to do anything. When we will test, we just deploy this and that's it. 😉

jcpunk commented on opmon-protobuf-dbwriter/opmon-dbwriter-deployment.yaml:
```diff
@@ -0,0 +1,80 @@
+      - image: ghcr.io/dune-daq/microservices:mroda-opmon_protobuf   ## TBC
+        imagePullPolicy: Always
+        name: opmon-protobuf-replica 
+        env:
```
Should this be updated to include the user/pass values?

mroda88 replied:
But this is not there yet, right? I added the option in the python, but the cluster still don't have that implemented. Until then there is not much we can do. Unless you want to change this while we add the microservice.

mroda88 committed:
add user pass for influx

mroda88 committed:
Update dependency, ers version updade. The change is not related to any use in the microservice, but it would be odd to keep using an old version

jcpunk commented on opmon-protobuf-dbwriter/entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+	--subscriber-topic $OPMON_DBWRITER_TOPIC \
+	--influxdb-address $OPMON_DBWRITER_INFLUX_HOST --influxdb-port $OPMON_DBWRITER_INFLUX_PORT \
+	--influxdb-name $OPMON_DBWRITER_TABLE --influxdb-timeout $OPMON_DBWRITER_BATCH_SIZE_MS \
+	--debug False --influxdb-create True --influxdb-username $OPMON_DBWRITER_INFLUX_USER --influxdb-password OPMON_DBWRITER_INFLUX_PASSWORD 
```
`OPMON_DBWRITER_INFLUX_PASSWORD` is missing the `$`

mroda88 replied:
Fixed, pushed and image recreated.

mroda88 committed:
fix mistake

mroda88 committed:
Fix ordering of variables

jcpunk commented on opmon-protobuf-dbwriter/entrypoint.sh:
```diff
@@ -0,0 +1,14 @@
+	--subscriber-topic $OPMON_DBWRITER_TOPIC \
+	--influxdb-address $OPMON_DBWRITER_INFLUX_HOST --influxdb-port $OPMON_DBWRITER_INFLUX_PORT \
+	--influxdb-name $OPMON_DBWRITER_TABLE --influxdb-timeout $OPMON_DBWRITER_BATCH_SIZE_MS \
+	--debug False --influxdb-create True --influxdb-username $OPMON_DBWRITER_INFLUX_USER --influxdb-password $OPMON_DBWRITER_INFLUX_PASSWORD 
```
Terribly minor nit, could you put `--debug False` on its own line at the bottom?

jcpunk commented on opmon-protobuf-dbwriter/opmon-dbwriter-deployment.yaml:
```diff
@@ -0,0 +1,84 @@
+  replicas: 1
+  selector:
+    matchLabels:
+      app: opmon-protobuf-replica
```
I generally try to have `matchLabels` match up with the `template.metadata.labels`

mroda88 replied:
That I'm not sure how to address it, I made the names more uniform. Let me know if this makes sense.

mroda88 committed:
Better entry point formatting

mroda88 committed:
Better matchLabels

mroda88 committed:
Correct name in deployment

mroda88 committed:
Fix type

mroda88 committed:
Correct use of logging

mroda88 committed:
Correct queue timeout handling

mroda88 committed:
Adding influx exception

mroda88 committed:
Better exception handling

mroda88 committed:
Testing dict approach

mroda88 committed:
Cleanup

jcpunk commented on opmon-protobuf-dbwriter/opmon-dbwriter-deployment.yaml:
```diff
@@ -0,0 +1,84 @@
+    pod-security.kubernetes.io/enforce-version: latest
+    pod-security.kubernetes.io/warn: baseline
+    pod-security.kubernetes.io/warn-version: latest
+  name: ers
```
Shouldn't this be `opmon`?

mroda88 replied:
Indeed!

mroda88 committed:
Final decoration
